R/response_process.R

```diff
@@ -0,0 +1,190 @@
+#' Process a Google API response
+#'
+#' @description
+#' Intended primarily for internal use in client packages that provide
+#' high-level wrappers for users. Typically applied as the final step in this
+#' sequence of calls:
+#'   * Request prepared with [request_build()].
+#'   * Request made with [request_make()].
+#'   * Response processed with [response_process()].
+#'
+#' All that's needed for a successful request is to parse the JSON extracted via
+#' `httr::content()`. Therefore, the main point of `response_process()` is to
+#' handle less happy outcomes:
+#'   * Status code in the 100s (information) or 300s (redirection). These are
+#'     unexpected.
+#'   * Non-JSON content type, such as HTML.
+#'   * Status codes in the 400s (client error) and 500s (server error). The
+#'     structure of the error payload varies across Google APIs and we try to
+#'     create a useful message for all variants we know about.
+
+#'
+#' @details
+#' A redacted version of the `resp` input is returned in the condition (auth
+#' tokens are removed). Use functions such as `rlang::last_error()` or
+#' `rlang::catch_cnd()` to capture the condition and do a more detailed forensic
+#' examination.
+#'
+#' @param resp Object of class `response` from [httr].
+#' @param raw Logical. Whether to return parsed content (`FALSE`) or in `raw`
+#'   form.
+#'
+#' @return The content of the request.
+#' @family requests and responses
+#' @export
+response_process <- function(resp, raw = FALSE) {
+  if (httr::status_code(resp) == 204) {
+    return(TRUE)
+  }
+
+  check_response(resp)
+  content <- httr::content(resp, as = "raw")
+
+  if (raw) {
+    content
+  } else {
+    jsonlite::fromJSON(rawToChar(content), simplifyVector = FALSE)
+  }
+}
+
+check_response <- function(resp) {
+  code <- httr::status_code(resp)
+  # code could be in 100s, 200s, 300s, 400s, 500s
+
+  if (code < 200 || (code >= 300 && code < 400)) {
+    message <- c(
+      httr::http_status(resp)$message,
+      "  * HTTP status codes in the 100s and 300s are unexpected."
+    )
+    stop_request_failed(resp, message = message)
+  }
+  # code could be in 200s, 400s, 500s
+
+  if (code >= 200 && code < 300) {
+    check_for_json(resp)
+    return()
+  }
+  # code could be: 400-499, 500-599
+
+  google_error(resp)
+}
+
+stop_request_failed <- function(resp, ..., message = NULL, .subclass = NULL) {
```

Comment:
Everything routes through here.

Reply:
I think we should talk about this in person, because this feels like it falls in an uncanny valley.